tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/AvroKafkaSourceTest.java

```diff
@@ -0,0 +1,496 @@
+        log.info("creating schema registry kafka {}",  getBootstrapServersOnDockerNetwork());
+        this.schemaRegistryContainer = new SchemaRegistryContainer(getBootstrapServersOnDockerNetwork());
+        cluster.startService(schemaRegistryContainerName, schemaRegistryContainer);
+        sourceConfig.put("bootstrapServers", getBootstrapServersOnDockerNetwork());
+        sourceConfig.put("groupId", "test-source-group");
+        sourceConfig.put("fetchMinBytes", 1L);
+        sourceConfig.put("autoCommitIntervalMs", 10L);
+        sourceConfig.put("sessionTimeoutMs", 10000L);
+        sourceConfig.put("heartbeatIntervalMs", 5000L);
+        sourceConfig.put("topic", kafkaTopicName);
+        sourceConfig.put("consumerConfigProperties",
+                ImmutableMap.of("schema.registry.url", getRegistryAddressInDockerNetwork())
+        );
+    }
+
+    private class EnhancedKafkaContainer extends KafkaContainer {
+
+        public EnhancedKafkaContainer(DockerImageName dockerImageName) {
+            super(dockerImageName);
+        }
+
+        @Override
+        public String getBootstrapServers() {
+            // we have to override this function
+            // because we want the Kafka Broker to advertise itself
+            // with the docker network address
+            // otherwise the Kafka Schema Registry won't work
+            return "PLAINTEXT://" + kafkaContainerName + ":9093";
+        }
+
+    }
+
+    protected EnhancedKafkaContainer createKafkaContainer(PulsarCluster cluster) {
+        return (EnhancedKafkaContainer) new EnhancedKafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:6.0.1"))
+                .withEmbeddedZookeeper()
+                .withCreateContainerCmdModifier(createContainerCmd -> createContainerCmd
+                        .withName(kafkaContainerName)
+                );
+    }
+
+    public void stopKafkaContainers(PulsarCluster cluster) {
+        if (null != schemaRegistryContainer) {
+            cluster.stopService(schemaRegistryContainerName, schemaRegistryContainer);
+        }
+        if (null != kafkaContainer) {
+            cluster.stopService(kafkaContainerName, kafkaContainer);
+        }
+    }
+
+    public void prepareSource() throws Exception {
+        log.info("creating topic");
+        ExecResult execResult = kafkaContainer.execInContainer(
+            "/usr/bin/kafka-topics",
+            "--create",
+            "--zookeeper",
+                getZooKeeperAddressInDockerNetwork(),
+            "--partitions",
+            "1",
+            "--replication-factor",
+            "1",
+            "--topic",
+            kafkaTopicName);
+        assertTrue(
+            execResult.getStdout().contains("Created topic"),
+            execResult.getStdout());
+
+    }
+
+    private String getZooKeeperAddressInDockerNetwork() {
+        return kafkaContainerName +":2181";
+    }
+
+    private <T extends GenericContainer> void testSource()  throws Exception {
```

Comment:
it the type parameter (`<T extends GenericContainer>`) required?

Reply:
it is a left over. I will remove it, thanks

Reply:
done